### PR TITLE
Restructure plotting

### PR DESCRIPTION
**Pull request by ahartikainen**
Restructure plotting.

backend related defaults to backend code.

### CODE REVIEW

**Comment by canyon289** [comment]
@ahartikainen Do you need help with this? Seems like theres a lot going on

**Comment by ahartikainen** [comment]
Sure if you have time.

I will continue when I have some time.

Basically there is couple of lines in each main plot function that is moved to backend functions, and then some of the inner kw are changed.

Then there is still a lot to do. I will add discussion here when the time is right.

**Comment by canyon289** [comment]
Wow @ahartikainen you did it. Is this still in draft stage? I'm dying to hit that merge button

**Comment by ahartikainen** [comment]
It still in draft. There are some normalization to be done

**Comment by ahartikainen** [comment]
Ready for review.

This PR enables mpl=>3.3 and doesn't change the user-interface.

File: arviz/plots/backends/matplotlib/loopitplot.py [hunk 1]
```diff
@@ -40,6 +50,55 @@ def plot_loo_pit(
...
     }
+
+    (figsize, _, _, xt_labelsize, linewidth, _) = _scale_fig_size(figsize, textsize, 1, 1)
+
+    plot_kwargs = matplotlib_kwarg_dealiaser(plot_kwargs, "plot")
+    plot_kwargs["color"] = to_hex(color)
+    plot_kwargs.setdefault("linewidth", linewidth * 1.4)
+    if isinstance(y, str):
+        label = ("{} LOO-PIT ECDF" if ecdf else "{} LOO-PIT").format(y)
+    elif isinstance(y, DataArray):
```
Comment (OriolAbril) [inline arviz/plots/backends/matplotlib/loopitplot.py:61]:
```suggestion
    elif isinstance(y, DataArray) and y.name is not None:
```

File: arviz/plots/backends/matplotlib/loopitplot.py [hunk 2]
```diff
@@ -40,6 +50,55 @@ def plot_loo_pit(
...
+        label = ("{} LOO-PIT ECDF" if ecdf else "{} LOO-PIT").format(y.name)
+    elif isinstance(y_hat, str):
+        label = ("{} LOO-PIT ECDF" if ecdf else "{} LOO-PIT").format(y_hat)
+    elif isinstance(y_hat, DataArray):
```
Comment (OriolAbril) [inline arviz/plots/backends/matplotlib/loopitplot.py:65]:
```suggestion
    elif isinstance(y_hat, DataArray) and y_hat.name is not None:
```

File: arviz/plots/backends/matplotlib/ppcplot.py [hunk 3]
```diff
@@ -29,21 +30,38 @@ def plot_ppc(
...
+            alpha = 1
+        else:
+            if kind.lower() == "scatter":
+                alpha = 0.7
+            else:
+                alpha = 0.2
+
+    if jitter is None:
+        jitter = 0.0
+    assert jitter >= 0.0
```
Comment (OriolAbril) [inline arviz/plots/backends/matplotlib/ppcplot.py:63]:
```suggestion
    if jitter < 0.0:
        raise ValueError("jitter must be >=0")
```

File: arviz/plots/backends/matplotlib/traceplot.py [hunk 4]
```diff
@@ -125,6 +133,61 @@ def plot_trace(
...
+    hist_kwargs.setdefault("alpha", 0.35)
+
+    if plot_kwargs is None:
+        plot_kwargs = {}
+    if fill_kwargs is None:
+        fill_kwargs = {}
+    if rug_kwargs is None:
+        rug_kwargs = {}
+    if rank_kwargs is None:
+        rank_kwargs = {}
```
Comment (OriolAbril) [inline arviz/plots/backends/matplotlib/traceplot.py:189]: I think these should be `kwargs = dealiaser(...`

File: arviz/plots/bpvplot.py [hunk 5]
```diff
@@ -22,7 +20,7 @@ def plot_bpv(
     reference="samples",
     n_ref=100,
     hdi_prob=0.94,
-    color="C0",
+    color="blue",
```
Comment (OriolAbril) [inline arviz/plots/bpvplot.py:23]: why blue? isn't it passed to to_hex? I prefer C0 as a default

File: arviz/plots/energyplot.py [hunk 6]
```diff
@@ -97,8 +97,6 @@ def plot_energy(
     types = "hist" if kind in {"hist", "histogram"} else "plot"
     plot_kwargs = matplotlib_kwarg_dealiaser(plot_kwargs, types)
```
Comment (OriolAbril) [inline arviz/plots/energyplot.py:98]: should this still be here?

File: arviz/plots/backends/bokeh/bpvplot.py [hunk 7]
```diff
@@ -47,6 +47,22 @@ def plot_bpv(
         **backend_kwarg_defaults(("dpi", "plot.bokeh.figure.dpi"),),
         **backend_kwargs,
     }
+
+    to_hex(color)
```
[the remark this answers is not in the document]
Comment (ahartikainen) [reply arviz/plots/backends/bokeh/bpvplot.py:51]: Yes I think it should be

File: arviz/plots/backends/bokeh/bpvplot.py [hunk 8]
```diff
@@ -47,6 +47,22 @@ def plot_bpv(
...
+
+    if plot_ref_kwargs is None:
+        plot_ref_kwargs = {}
+    if kind == "p_value" and reference == "analytical":
+        plot_ref_kwargs.setdefault("line_color", "black")
+        plot_ref_kwargs.setdefault("line_dash", "dashed")
+    else:
+        plot_ref_kwargs.setdefault("alpha", 0.1)
```
[the remark this answers is not in the document]
Comment (ahartikainen) [reply arviz/plots/backends/bokeh/bpvplot.py:59]: bokeh kwargs are currently a mess, so I didn't touch them for now.

Also most bokeh kwargs are currently called explicitly

File: arviz/plots/backends/bokeh/bpvplot.py [hunk 9]
```diff
@@ -2,13 +2,15 @@
 import numpy as np
 from bokeh.models import BoxAnnotation
 from bokeh.models.annotations import Title
+from matplotlib.colors import to_hex
```
[the remark this answers is not in the document]
Comment (ahartikainen) [reply arviz/plots/backends/bokeh/bpvplot.py:5]: For bokeh, I think we should move closer ton"native" solution and use bokeh related tools

File: arviz/plots/backends/bokeh/distplot.py [hunk 10]
```diff
@@ -44,6 +44,16 @@ def plot_dist(
         **backend_kwarg_defaults(),
         **backend_kwargs,
     }
+
+    hist_kwargs = {} if hist_kwargs is None else hist_kwargs
+    if kind == "hist":
+        hist_kwargs.setdefault("cumulative", cumulative)
+        hist_kwargs.setdefault("fill_color", color)
+        hist_kwargs.setdefault("line_color", color)
```
[the remark this answers is not in the document]
Comment (ahartikainen) [reply arviz/plots/backends/bokeh/distplot.py:52]: probably

File: arviz/plots/backends/bokeh/mcseplot.py [hunk 11]
```diff
@@ -76,7 +86,7 @@ def plot_mcse(
                 hline_mean = Span(
                     location=mean_mcse_i,
                     dimension="width",
-                    line_color="black",
+                    line_color=extra_kwargs["color"],
                     line_width=extra_kwargs["linewidth"] * 2,
                     line_alpha=extra_kwargs["alpha"],
```
[the remark this answers is not in the document]
Comment (ahartikainen) [reply arviz/plots/backends/bokeh/mcseplot.py:91]: Yes, we need to clean these. I have an idea for scheme how to handle these

**Comment by ahartikainen** [comment]
@OriolAbril @canyon289 I'm starting to think that maybe we should move closer to "native" bokeh settings and colors.

This would mean that we don't change "red" to mpl red but use bokeh red. We could have handling for special colors

    C0-C9 --> let's fix this by our selves
    `red_matplotlib` or `red_mpl` --> force mpl colors?

File: arviz/plots/backends/bokeh/bpvplot.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (OriolAbril) [reply arviz/plots/backends/bokeh/bpvplot.py:59]: Let's leave it like this then

**Comment by OriolAbril** [comment]
For base colors, both libraries use CSS colors so `red` is exactly the same color for mpl and for bokeh. https://docs.bokeh.org/en/latest/docs/reference/colors.html#bokeh-colors-named and https://matplotlib.org/3.1.0/gallery/color/named_colors.html

The conversion using mpl helpers has the advantage that then, in addition to CSS colors, all the following are also supported:
* C0... so that defaults easily mute when changing the stylesheet
* single letter colors 
* tableau colors
* xkcd colors

**Comment by ahartikainen** [comment]
Ok, if the colors are the same, I'm fine with the helper function.

**Comment by canyon289** [comment]
I'll take a look at this in next ~24 hours but dont let me block a merge either

**Comment by canyon289** [comment]
I dont want there to be a merge conflict so I'm going to merge this and keep reviewing. If stuff comes up I'll make an issue ticket

**Pull request merged**